proposed/2021/CentralPackageManagementTransitive.md

```diff
@@ -0,0 +1,201 @@
+
+**`Directory.Packages.props`**
+```xml
+<Project>
+  <ItemGroup>
+    <PackageVersion Include="PackageA" Version="1.0.0" />
+    <PackageVersion Include="PackageB" Version="2.0.0" />
+  </ItemGroup>
+</Project>
+```
+
+All projects in the repository that reference `PackageA` or `PackageB` explicitly will use the specified value in a corresponding `PackageVersion` item.
+
+If a user wants to override the transitive version of `PackageC`, they must add an explicit `PackageReference` item:
+```diff
+<ItemGroup>
+  <PackageReference Include="PackageA" />
+  <PackageReference Include="PackageB" />
++ <PackageReference Include="PackageC" />
+</ItemGroup>
+```
+
+And add a corresponding `PackageVersion` item to Directory.Packages.props:
+```diff
+<ItemGroup>
+  <PackageVersion Include="PackageA" Version="1.0.0" />
+  <PackageVersion Include="PackageB" Version="2.0.0" />
++ <PackageVersion Include="PackageC" Version="5.0.0" />
+</ItemGroup>
+```
+
+The result is the same as before where the resolved `PackageC` is version **`5.0.0`** and affects the restored packages, reference assemblies passed to the compiler, files copied to the output directory, and dependencies in used during package creation.  The transitive dependency is no longer considered transitive and instead is an explicit dependency.
+
+### Proposal to allow transitive version selection based on specified PackageVersion items
+In the previous example, a user must specify an explicit `PackageReference` to a package to override the transitive version.  This elevates the reference from transitive to explicit and behaves as a user expects.  However, some users want a `PackageVersion` to automatically override a transitive version even without an explicit top-level `PackageReference`.  The reason for this feature request is for better **scale** and **performance**.
+
+#### Addressing Scale
+For larger repositories, requiring users to specify explicit `PackageReference` items in each project can become difficult.  This is because you may need to add it to dozens of projects in order to unify the version used across the repo.  Since CPVM is used to centralize package versions, it makes sense to have transitive reference version overrides to be implied when specifying a `PackageVersion`.  There are however at least two drawbacks of this approach:
+
+1. **Confusion** about "pinning" a transitive version - Many users consider overriding a transitive version to be pinning that version in the graph.  In some cases they consider this gesture only part of Restore and Build.  But since they've overridden a transitive version, this also flows to Pack which affects the versions used as dependencies.  When transitive versions are are "pinned", they are really just elevated to top-level dependencies which causes this confusion.  We will need to educate users on how this feature works so they understand that it is not really just pinning a version but instead is actually making it an implicit reference.
```

Comment:
Should we explain a bit **why** a pinned package is elevated to a direct dependency?  A brief example of how if this was not the case, the package author might accidentally use an API only available in the higher version of the transitive package, and then when another developer uses the package author's package in their own app, the library crashes due to missing method exceptions, or similar.

Reply:
Does line 65 cover it?

> Note that since **`PackageC` is no longer considered transitive, it must now be elevated to be a top-level dependency of my package**.  This is because `ClassLibrary1` was built against version `5.0.0` of `PackageC` so the transitive dependencies of `PackageA` and `PackageB` are no longer accurate.  The act of "pinning" a version of a transitive dependency effectively transformed it into an explicit dependency.